{"cells":[{"metadata":{},"cell_type":"markdown","source":"# Information Entry Script for Metagenomic Analysis\nThis script allows you to enter in information into the Jupyter_metagenomic_analysis.ipynb file. This script will prompt you to enter in relevant information as well as locate necessary files for the analysis, and this information will be stored into a .csv file that can be read in by the Jupyter_metagenomic_analysis.ipynb file. For output verification, you can open this .csv file (named --.csv) to ensure the information looks correct before running the analysis script. \n\nWhen running this script, run all of the cells sequentially one after another after you are done editing the cell preceding it. This will ensure the script runs smoothly."},{"metadata":{},"cell_type":"markdown","source":"For questions about this script, please contact Akhil Gupta (gupta305@umn.edu). "},{"metadata":{},"cell_type":"markdown","source":"---"},{"metadata":{},"cell_type":"markdown","source":"## Import UI Elements Script\nThis will load in all necessary widgets. Ignore the warning message for now"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Importing necessary functions from another .ipynb script\n%run ui_elem.ipynb","execution_count":1,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"----------------"},{"metadata":{},"cell_type":"markdown","source":"## Text Entry\nHere you should specify the column name (needs a better description) as well as give a name for the folder where the graphs and statistics should be stored. "},{"metadata":{"trusted":true},"cell_type":"code","source":"direct_name_data","execution_count":2,"outputs":[{"output_type":"display_data","data":{"text/plain":"VBox(children=(Text(value='ID', description='Column Name:', layout=Layout(width='70%'), placeholder='What is t…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"eac82beb8fc349a4b656460654233a81"}},"metadata":{}}]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Test to see if the value in the text boxes are saved\nprint(\"Sample column ID: \" + sample_column_id.value)\nprint(\"Graph Output Directory Name: \" + graph_output_dir.value)\nprint(\"Stats Output Directory Name: \" + stats_output_dir.value)","execution_count":3,"outputs":[{"output_type":"stream","text":"Sample column ID: ID\nGraph Output Directory Name: graphs\nStats Output Directory Name: stats\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"--------------------------------------------------"},{"metadata":{},"cell_type":"markdown","source":"## File Paths\n\nPrior to running this portion of the script, be sure that all the files that will be used for analysis are uploaded to the data folder in the file directory where this script was opened from. Open the directory named \"data\", and upload all files that will be necessary to run the analysis. All the scripts that are required for the analysis are listed below. \n\nOnce the files have been uploaded, click on the circulary icon in the top left of the screen to restart the python kernal - Jupyter doesn't update the list below while the kernel is running, you will instead have to restart the kernel. Once restarted, the file should appear in the drop down menu below, where you can then select the corresponding file. The current placeholders will work just fine if you don't have a certain file on hand. \n\n__*This section needs more work in terms of descriptions*__"},{"metadata":{},"cell_type":"markdown","source":"#### Loading in resistome files"},{"metadata":{"trusted":true},"cell_type":"code","source":"resistome_filenames","execution_count":4,"outputs":[{"output_type":"display_data","data":{"text/plain":"VBox(children=(Dropdown(description='AMR Count Matrix File:', index=7, layout=Layout(width='50%'), options=('t…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"0197e99407ef44a6927d45e63fa6eef4"}},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"#### Loading in microbiome files"},{"metadata":{"scrolled":true,"trusted":true},"cell_type":"code","source":"microbiome_filenames","execution_count":5,"outputs":[{"output_type":"display_data","data":{"text/plain":"VBox(children=(Dropdown(description='Biom File:', index=4, layout=Layout(width='50%'), options=('test_16S_dna-…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"f6e6bcee64eb4327b8794a7a5c852a04"}},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"#### Save the file names\nThis will also print the output so you can double check its accuracy"},{"metadata":{"trusted":true},"cell_type":"code","source":"save_filepath_button","execution_count":6,"outputs":[{"output_type":"display_data","data":{"text/plain":"Button(description='Save the filepaths for analysis', icon='save', layout=Layout(width='70%'), style=ButtonSty…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"c78ed7edb6b944e697ba89b4a4c07fea"}},"metadata":{}},{"output_type":"stream","text":"AMR Count Matrix Filepath: test_AMR_analytic_matrix.csv\nAMR Metadata Filepath: test_AMR_metadata.csv\nMEGARes Annotation Filepath: test_AMR_megares_annotations_v1.03.csv\nBiom Filepath: test_16S_otu_table_json.biom\nFasta Filepath: test_16S_tree.nwk\nTaxonomy Filepath: test_16S_taxonomy.tsv\nMicrobiome Temp Metadata Filepath: test_16S_metadata.csv\n\nAll filepaths saved\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"---"},{"metadata":{},"cell_type":"markdown","source":"## Exploratory Variables\nThe following will allow you to enter in variables for the analyses. It's based on variables in your metadata.csv file that you want to use for EXPLORATORY analysis (NMDS, PCA, alpha rarefaction, barplots)"},{"metadata":{},"cell_type":"markdown","source":"### Sliders\nThe slider allows you to choose how many separate analyses should be run, and how many of them should be AMR analyses and how many should be Microbiome. \n\n\n### Explanatory Variables\nEvery time the slider value is changed, the code to generate the boxes should be rerun in order to update them. They will automatically contain the correct number of boxes based on the values entered in the slider. \n\n**Name**: \n\n**Subset**: Subset will allow you to filter out variables of interest and remove unnecessary variables. To select variables of interest, the format should be “*column-2 == column-variable-1*”. This is exactly how you might enter this into R. \n\nTo remove certain variables from the analysis, the format should be “*column_2 != column_variable_2*”. The key point is having the *!* symbol instead of the first exclamation point.\n\n**Explanatory Variable**: This should also be the name of a column of interest for analysis. \n\n*NOTE*: Exploratory variables cannot be numeric. \n\n**Order**: This should describe the order that will be used during the analysis and when printing out result plots. *Each item in the list should be separated by a comma.*"},{"metadata":{"trusted":true},"cell_type":"code","source":"display(exp_graph_var_amr)\ndisplay(exp_graph_var_microbiome)","execution_count":7,"outputs":[{"output_type":"display_data","data":{"text/plain":"IntSlider(value=5, continuous_update=False, description='AMR', max=10)","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"12e81122da154619a203e74057520d7b"}},"metadata":{}},{"output_type":"display_data","data":{"text/plain":"IntSlider(value=5, continuous_update=False, description='Microbiome', max=10)","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"db9deef4e7bc4bd9917959cf41105bef"}},"metadata":{}}]},{"metadata":{"scrolled":false,"trusted":true},"cell_type":"code","source":"var_info(exp_graph_var_amr.value, exp_graph_var_microbiome.value)","execution_count":8,"outputs":[{"output_type":"display_data","data":{"text/plain":"Tab(children=(Accordion(children=(VBox(children=(Text(value='', layout=Layout(width='70%'), placeholder='name'…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"9f7024341ac045838ac9580bb59b1d45"}},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"--------"},{"metadata":{},"cell_type":"markdown","source":"## IMPORTANT STEP\nMake sure to run the code below once you have finished entering the data into the boxes above. This will ensure that the data is stored correctly and will be output into the .csv file. "},{"metadata":{"trusted":true},"cell_type":"code","source":"# Saves and prints the variables entered above into a list to be used when creating the .csv file\nlist_vals_a, list_vals_m = save_print_variables(exp_graph_var_amr.value, exp_graph_var_microbiome.value)","execution_count":16,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"---"},{"metadata":{},"cell_type":"markdown","source":"## Outputting information into .csv file\nThis below will now store everything entered above into a .csv file that can be read in by the analysis script. The .csv file will be stored in the current working directory where this script is also stored. If your analysis script is located in another directory, be sure to move the .csv file into that same directory before running that analysis script. "},{"metadata":{"trusted":true},"cell_type":"code","source":"display(vars_save_button)\nvars_save_button.on_click(vars_to_csv)","execution_count":10,"outputs":[{"output_type":"display_data","data":{"text/plain":"Button(description='Save variables for analysis script', icon='save', layout=Layout(width='50%'), style=Button…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"199dce50cd1440f78f0a99c1bb2e2b5f"}},"metadata":{}},{"output_type":"stream","text":"Variables Exported. Check directory for .csv file\nVariables Exported. Check directory for .csv file\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"---"},{"metadata":{},"cell_type":"markdown","source":"# Next Steps\n\nThe results from this script will be output in a .csv file that the R script will then use to run the rest of the analysis. You don't need to worry about the .csv file, all you need to do now is open the script named \"Jupyter_metagenomic_analysis.ipynb\" in the directory where this file was also located and run it. "},{"metadata":{},"cell_type":"markdown","source":"## Run staging script with R magic"},{"metadata":{"trusted":true},"cell_type":"code","source":"import rpy2.rinterface\n%load_ext rpy2.ipython","execution_count":17,"outputs":[{"output_type":"stream","text":"The rpy2.ipython extension is already loaded. To reload it, use:\n  %reload_ext rpy2.ipython\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"%%R -o AMR_melted_analytic -o microbiome_melted_analytic\n\n# I basically turned your jupyter notebook \"Jupyter_metagenomic_analysis.ipynb\" into the staging script.\nsource(\"staging_script.R\")","execution_count":null,"outputs":[{"output_type":"stream","text":"From cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\nFrom cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\nFrom cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\nFrom cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\nFrom cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\nFrom cffi callback <function _processevents at 0x7f76c49badd0>:\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/rpy2/rinterface_lib/callbacks.py\", line 264, in _processevents\n    @ffi_proxy.callback(ffi_proxy._processevents_def,\nKeyboardInterrupt\n","name":"stderr"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"AMR_melted_analytic","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#%%R\n\n# This part takes longer, so for now we'll keep it seperate.\n## Run code to make some exploratory figures, zero inflated gaussian model, and output count matrices.\n#suppressMessages(source('scripts/print_figures.R'))","execution_count":22,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Example of python widget"},{"metadata":{"trusted":true},"cell_type":"code","source":"import ipywidgets as widgets\nfrom ipywidgets import interact, interact_manual\nimport os\nfrom IPython.display import Image","execution_count":20,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## The example below has to be fixed because I had to hard code the directory path\nYou'll probably have an easier time editing this than I was having.\n\nWe can either make one widget per dataset type (AMR/Microbiome), or modify the widget to have another button that lets you select the dataset type first, then the exploratory variable, and then the figure to see"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Create microbiome widgets\ndirectory = widgets.Dropdown(options=os.listdir('graphs/Microbiome/'))\nimages = widgets.Dropdown(options=os.listdir('graphs/Microbiome/' + directory.value))\n\n# Updates the image options based on directory value\ndef update_images(*args):\n    images.options = os.listdir(directory.value)\n\n# Tie the image options to directory value\ndirectory.observe(update_images, 'value')\n\n# Show the images\ndef show_images(fdir, file):\n    display(Image(f'{fdir}/{file}'))\n\n_ = interact(show_images, fdir=('graphs/Microbiome/' + directory.value), file=images)","execution_count":21,"outputs":[{"output_type":"display_data","data":{"text/plain":"interactive(children=(Text(value='graphs/Microbiome/Treatment', description='fdir'), Dropdown(description='fil…","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"84ac5260087b4b77b4bdc917271004c8"}},"metadata":{}}]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}