{"cells":[{"metadata":{},"cell_type":"markdown","source":"## Setup"},{"metadata":{"scrolled":false,"trusted":false},"cell_type":"code","source":"import ipywidgets as widgets\nfrom ipywidgets import Button, Layout, IntSlider, HBox, VBox, interact, interact_manual\nimport os\nimport csv\nfrom IPython.display import display, clear_output, Markdown, Image\nimport rpy2.rinterface\nimport copy\n%load_ext rpy2.ipython","execution_count":1,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"import pandas as pd","execution_count":1,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## Initializing all relevant variables as global variables\nglobal sample_column_id\nglobal graph_output_dir\nglobal stats_output_dir\nglobal amr_count_matrix_filepath\nglobal amr_metadata_filepath\nglobal megares_annotation_filename\nglobal biom_file\nglobal tre_file\nglobal tax_fasta\nglobal taxa_file\nglobal microbiome_temp_metadata_file\nglobal list_vals_a\nglobal list_vals_m","execution_count":2,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### View metadata file and select column ID"},{"metadata":{"trusted":false},"cell_type":"code","source":"data_f = os.listdir(str(os.getcwd()) + \"/data\")\ndata_f_csv = [x for x in data_f if \".csv\" in x]\n\nmeta_filename = widgets.Dropdown(\n    options=data_f_csv,\n    description='Select CSV File',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Show metadata head button\ncol_names_meta = []\ndef open_meta(b):\n    global meta\n    meta = pd.read_csv(str(\"data/\" + meta_filename.value))\n    display(meta.head())\n    col_names_meta = list(meta.columns)\n\nshow_meta_button = widgets.Button(\n    description=\"Print head of metadata file\", \n    icon = 'open',\n    layout=Layout(width='70%'))\n\nshow_meta_button.on_click(open_meta)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"def select_col():\n    global sample_column_id\n    sample_column_id = widgets.Dropdown(\n        options=list(meta.columns),\n        description='Column Names of Selected File',\n        disabled=False,\n        layout=Layout(width='50%'),\n        style = {'description_width': 'initial'}\n    )\n    display(sample_column_id)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Text Entry Widgets"},{"metadata":{"trusted":false},"cell_type":"code","source":"graph_output_dir = widgets.Text(\n    value='graphs',\n    placeholder='Name the output directory for where the graphs should be saved',\n    description='Directory Name for Graphs: ',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n\nstats_output_dir = widgets.Text(\n    value='stats',\n    placeholder='Name the output directory for where the statistics should be saved',\n    description='Directory Name for Stats: ',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n\n# direct_name_data = VBox([sample_column_id, graph_output_dir, stats_output_dir])","execution_count":3,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## File Paths\nCreating elements to enter in file names in text boxes"},{"metadata":{"trusted":false},"cell_type":"code","source":"amr_count_file = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_AMR_analytic_matrix.csv\",\n    description='AMR Count Matrix File:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\namr_metadata_file = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_AMR_metadata.csv\",\n    description='Metadata File:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\nmegares_annot = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_AMR_megares_annotations_v1.03.csv\",\n    description='MEGARes Annotation File:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\nresistome_filenames = VBox([amr_count_file, amr_metadata_file, megares_annot])","execution_count":4,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"biom = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_16S_otu_table_json.biom\",\n    description='Biom File:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\ntre = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_16S_tree.nwk\",\n    description='tree file:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\ntax = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_16S_dna-sequences.fasta\",\n    description='fasta file:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\ntaxa = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_16S_taxonomy.tsv\",\n    description='taxonomy file:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\nmicrobiome_temp = widgets.Dropdown(\n    options=os.listdir(str(os.getcwd()) + \"/data\"),\n    value = \"test_16S_metadata.csv\",\n    description='microbiome temp metadata file:',\n    disabled=False,\n    layout=Layout(width='50%'),\n    style = {'description_width': 'initial'}\n)\n\nmicrobiome_filenames = VBox([biom, tre, tax, taxa, microbiome_temp])","execution_count":5,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## AMR Text Boxes\n\nmicrobiome_filenames = VBox([biom, tre, tax, taxa, microbiome_temp])","execution_count":6,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## Function to save all the filepaths to the normal variable expected by the rest of the code\ndef save_filepaths(b):\n    \n    global amr_count_matrix_filepath\n    global amr_metadata_filepath\n    global megares_annotation_filename\n    \n    global biom_file\n    global tre_file\n    global tax_fasta\n    global taxa_file\n    global microbiome_temp_metadata_file\n    \n    amr_count_matrix_filepath = \"data/\" + str(amr_count_file.value)\n    amr_metadata_filepath = \"data/\" + str(amr_metadata_file.value)\n    megares_annotation_filename = \"data/\" + str(megares_annot.value)\n    \n    biom_file = \"data/\" + str(biom.value)\n    tre_file = \"data/\" + str(tre.value)\n    tax_fasta = \"data/\" + str(tax.value)\n    taxa_file = \"data/\" + str(taxa.value)\n    microbiome_temp_metadata_file = \"data/\" + str(microbiome_temp.value)\n    \n    print(\"AMR Count Matrix Filepath: \" + str(amr_count_file.value))\n    print(\"AMR Metadata Filepath: \" + str(amr_metadata_file.value))\n    print(\"MEGARes Annotation Filepath: \" + str(megares_annot.value))\n    print(\"Biom Filepath: \" + str(biom.value))\n    print(\"Fasta Filepath: \" + str(tre.value))\n    print(\"Taxonomy Filepath: \" + str(taxa.value))\n    print(\"Microbiome Temp Metadata Filepath: \" + str(microbiome_temp.value))\n    print()\n    print(\"All filepaths saved\")\n    \n","execution_count":7,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"save_filepath_button = widgets.Button(\n    description=\"Save the filepaths for analysis\", \n    icon = 'save',\n    layout=Layout(width='70%'))\n\nsave_filepath_button.on_click(save_filepaths)","execution_count":8,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## AMR Exploratory variables\nMultiple text input boxes backend code"},{"metadata":{"trusted":false},"cell_type":"code","source":"### Making the function for the slider to choose number of accordions \nexp_graph_var_amr = widgets.IntSlider( \n    value=5,\n    min=0,\n    max=10,\n    step=1,\n    description='AMR',\n    continuous_update=False,\n    orientation='horizontal',\n    readout=True,\n    readout_format='d'\n)\n\nexp_graph_var_microbiome = widgets.IntSlider( \n    value=5,\n    min=0,\n    max=10,\n    step=1,\n    description= 'Microbiome',\n    continuous_update=False,\n    orientation='horizontal',\n    readout=True,\n    readout_format='d'\n)","execution_count":9,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"### Making the different text boxes for the 4 variables\nname = widgets.Text(\n    value='',\n    placeholder='Name',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n\n\n#widgets.Text(value='', placeholder='Name', disabled=False, layout=Layout(width='70%'), style = {'description_width': 'initial'})\n\n\nsubsets = widgets.Text(\n    value='',\n    placeholder='Subsets',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n\nexploratory_var = widgets.Text(\n    value='',\n    placeholder='Exploratory Variable',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n\norder = widgets.Text(\n    value='',\n    placeholder='Order',\n    disabled=False,\n    layout=Layout(width='70%'),\n    style = {'description_width': 'initial'}\n)\n","execution_count":10,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"### Creating a vertical box to store all of the text boxes into a single object \nmenu1 = widgets.VBox([name, subsets, exploratory_var, order])","execution_count":11,"outputs":[]},{"metadata":{"scrolled":true,"trusted":false},"cell_type":"code","source":"### Updating the number of accordion pages based on the value selected in the slider above, then prints it\ndef graph_vars_amr(exp_graph_var):\n    list_widgets_a = []\n    \n    # This creates all the new variables to store the text boxes in\n    for i in range(exp_graph_var):\n        \n        name_box = \"global name_a{}; name_a{} = widgets.Text(placeholder='name',layout=Layout(width='70%'))\".format(i,i)\n        subset_box = \"global subset_a{}; subset_a{} = widgets.Text(placeholder='Subset',layout=Layout(width='70%'))\".format(i,i)\n        exp_box = \"global exploratory_a{}; exploratory_a{} = widgets.Text(placeholder='Exploratory Variable',layout=Layout(width='70%'))\".format(i,i)\n        order_box = \"global order_a{}; order_a{} = widgets.Text(placeholder='Order',layout=Layout(width='70%'))\".format(i,i)\n        exec(name_box)\n        exec(subset_box)\n        exec(exp_box)\n        exec(order_box)\n             \n    \n    # This will assign a new menu_ variable for each accordion that needs to be printed\n    for i in range(exp_graph_var):\n        string = \"menu_a{} = widgets.VBox([name_a{}, subset_a{}, exploratory_a{}, order_a{}])\".format(i, i, i, i, i)\n        exec(string)\n\n    \n    # This should append all the menu variables into the list_widgest list and pass it into the accordion widget\n    for i in range(0, exp_graph_var):\n        string = \"list_widgets_a.append(menu_a{})\".format(i)\n        exec(string)\n    \n    # Creates and displays the final accordion widget\n    accordion_a = widgets.Accordion(children=list_widgets_a)\n    return accordion_a\n\n\ndef graph_vars_mic(exp_graph_var):\n    list_widgets_m = []\n    # This creates all the new variables to store the text boxes in\n    for i in range(exp_graph_var):\n        \n        name_box_m = \"global name_m{}; name_m{} = widgets.Text(placeholder='name',layout=Layout(width='70%'))\".format(i,i)\n        subset_box_m = \"global subset_m{}; subset_m{} = widgets.Text(placeholder='Subset',layout=Layout(width='70%'))\".format(i,i)\n        exp_box_m = \"global exploratory_m{}; exploratory_m{} = widgets.Text(placeholder='Exploratory Variable',layout=Layout(width='70%'))\".format(i,i)\n        order_box_m = \"global order_m{}; order_m{} = widgets.Text(placeholder='Order',layout=Layout(width='70%'))\".format(i,i)\n        exec(name_box_m)\n        exec(subset_box_m)\n        exec(exp_box_m)\n        exec(order_box_m)\n             \n    \n    # This will assign a new menu_ variable for each accordion that needs to be printed\n    for i in range(exp_graph_var):\n        string = \"menu_m{} = widgets.VBox([name_m{}, subset_m{}, exploratory_m{}, order_m{}])\".format(i, i, i, i, i)\n        exec(string)\n    \n    # This should append all the menu variables into the list_widgest list and pass it into the accordion widget\n    for i in range(0, exp_graph_var):\n        string = \"list_widgets_m.append(menu_m{})\".format(i)\n        exec(string)\n    \n    # Creates and displays the final accordion widget\n    accordion_m = widgets.Accordion(children=list_widgets_m)\n    return accordion_m\n","execution_count":12,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## Store all the variables in amr_exp in order to be written out to a csv file\nglobal list_vals_a\nglobal list_vals_m\nlist_vals_a = []\nlist_vals_m = []\n\ndef save_print_variables(amr, mic):\n    list_vals_a = []\n    list_vals_m = []\n    \n    exp = [\"_a\", \"_m\"]\n    num = [amr, mic]\n    \n    for i in range(2):\n        for j in range(num[i]):\n            analysis = exp[i]\n            exec(\"order_new{}{} = order_format(order{}{}.value)\".format(analysis, j, analysis, j))\n            exec(\"subset_new{}{} = subset_format(subset{}{}.value)\".format(analysis, j, analysis, j))\n            string = 'list_vals{}.append([name{}{}.value, subset_new{}{}, exploratory{}{}.value, order_new{}{}])'.format(analysis, analysis, j, analysis, j, analysis, j, analysis, j)\n            exec(string)\n\n    #print(list_vals_a)\n    #print(\"\")\n    #print(list_vals_m)\n    \n    return list_vals_a, list_vals_m","execution_count":13,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Create button to save variables from exploratory variables\n# list_vals_a, list_vals_m = save_print_variables(exp_graph_var_amr.value, exp_graph_var_microbiome.value)\n\ndef save_explore_vars(b):\n    list_vals_a, list_vals_m = save_print_variables(exp_graph_var_amr.value, exp_graph_var_microbiome.value)\n    print(\"Exploratory variables saved\")\n\nsave_print_vars_button = widgets.Button(\n    description=\"Save the exploratory variables for analysis\", \n    icon = 'save',\n    layout=Layout(width='70%'))\n\nsave_print_vars_button.on_click(save_explore_vars)\n\n# display(save_print_vars_button)\n\n","execution_count":15,"outputs":[{"ename":"NameError","evalue":"name 'order_format' is not defined","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-15-082759241b6d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Create button to save variables from exploratory variables\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mlist_vals_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist_vals_m\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msave_print_variables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexp_graph_var_amr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexp_graph_var_microbiome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m<ipython-input-13-31b88c8569b9>\u001b[0m in \u001b[0;36msave_print_variables\u001b[0;34m(amr, mic)\u001b[0m\n\u001b[1;32m     15\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m             \u001b[0manalysis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m             \u001b[0mexec\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"order_new{}{} = order_format(order{}{}.value)\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m             \u001b[0mexec\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"subset_new{}{} = subset_format(subset{}{}.value)\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m             \u001b[0mstring\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'list_vals{}.append([name{}{}.value, subset_new{}{}, exploratory{}{}.value, order_new{}{}])'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m<string>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: name 'order_format' is not defined"]}]},{"metadata":{"trusted":false},"cell_type":"code","source":"## Create the tabs to enter in variable data.\ndef var_info(amr, mic):\n    tab = widgets.Tab()\n    if mic == 0:\n        tab_contents = [\"AMR\"]\n        children = [graph_vars_amr(amr)]\n        tab.set_title(0, \"AMR\")\n    elif amr == 0:\n        tab_contents = [\"Microbiome\"]\n        children = [graph_vars_mic(mic)]\n        tab.set_title(0, \"Microbiome\")\n    else: \n        tab_contents = [\"AMR\", \"Microbiome\"]\n        children = [graph_vars_amr(amr), graph_vars_mic(mic)]\n        tab.set_title(0, \"AMR\")\n        tab.set_title(1, \"Microbiome\")\n    \n    tab.children = children\n    tab.titles = tab_contents\n    return tab","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Reformats the order variable\ndef order_format(order_og):\n    # Splits the text into the different items in the list\n    order_list = order_og.split(\",\")\n    \n    # Removes the white spaces before and after each item\n    order_list = [item.strip() for item in order_list]\n\n    # Adds the quotation marks around each item\n    order_list = ['\"' + item + '\"' for item in order_list]\n    \n    # Removing unnecessary characters from string\n    order_list = 'c({})'.format(order_list)\n    order_list = order_list.replace(\"[\", \"\")\n    order_list = order_list.replace(\"]\", \"\")\n    order_list = order_list.replace(\"'\", \"\")\n    order_list = order_list.replace(\", \", \",\")\n    \n    if order_og == \"\":\n        order_list = \"\"\n    \n    return order_list\n\n# Reformats the subset variable\ndef subset_format(subset_og):\n    # Splits the text into the different items in the list\n    order_list = subset_og.split(\",\")\n    \n    # Removes the white spaces before and after each item\n    order_list = [item.strip() for item in order_list]\n\n    # Adds the quotation marks around each item\n    order_list = [\"'\" + item + \"'\" for item in order_list]\n    \n    # Removing unnecessary characters from string\n    order_list = 'list({})'.format(order_list)\n    order_list = order_list.replace(\"[\", \"\")\n    order_list = order_list.replace(\"]\", \"\")\n    order_list = order_list.replace('\"', \"\")\n    order_list = order_list.replace(\", \", \",\")\n    \n    if subset_og == \"\":\n        order_list = \"\"\n    \n    \n    return order_list\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## USER ENTERED GUI\n# Function to save all variables into .csv file\ndef vars_to_csv(b):\n    with open('metagenome_analysis_vars.csv','w', newline='') as f:  \n        writer = csv.writer(f)\n        # Column names\n        writer.writerow(['sample_column_id', sample_column_id.value])\n        writer.writerow(['graph_output_dir', graph_output_dir.value])\n        writer.writerow(['stats_output_dir', stats_output_dir.value])\n\n        # Filepaths\n        writer.writerow(['amr_count_matrix_filepath', amr_count_matrix_filepath])\n        writer.writerow(['amr_metadata_filepath', amr_metadata_filepath])\n        writer.writerow(['megares_annotation_filename', megares_annotation_filename])\n        writer.writerow(['biom_file', biom_file])\n        writer.writerow(['tre_file', tre_file])\n        writer.writerow(['tax_fasta', tax_fasta])\n        writer.writerow(['taxa_file', taxa_file])\n        writer.writerow(['microbiome_temp_metadata_file', microbiome_temp_metadata_file])\n\n        # AMR exploratory variables\n        writer.writerow([\"AMR_exploratory_analyses\"])\n        writer.writerows(list_vals_a)\n\n        # Microbiome exploratory variables\n        writer.writerow([\"microbiome_exploratory_analyses\"])\n        writer.writerows(list_vals_m)\n    print(\"CUSTOM variables Exported. Check directory for .csv file\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"## DEFAULT INFORMATION\ndef vars_to_csv_default(b):\n    os.popen('cp metagenome_analysis_vars_DEFAULT.csv metagenome_analysis_vars.csv')\n    print(\"DEFAULT variables Exported. Check directory for .csv file\")\n    ","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"vars_save_button = widgets.Button(\n    description=\"Save CUSTOM variables for analysis script\", \n    icon = 'save',\n    layout=Layout(width='50%'))\n\nvars_save_button_default = widgets.Button(\n    description=\"Save DEFAULT variables for analysis script\", \n    icon = 'save',\n    layout=Layout(width='50%'))\n\nvars_save_button.on_click(vars_to_csv)\nvars_save_button_default.on_click(vars_to_csv_default)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## R Code"},{"metadata":{"trusted":false},"cell_type":"code","source":"def stage_script(b):\n    %R -o amr_melted_analytic -o microbiome_melted_analytic -o amr_melted_raw_analytic -o microbiome_melted_raw_analytic\n    %R source(\"staging_script.R\")\n\nstage_script_button = widgets.Button(\n    description=\"Run the Staging Script to initialize objects\", \n    icon = 'gear',\n    layout=Layout(width='50%'))\n\nstage_script_button.on_click(stage_script)\n\n# display(stage_script_button)","execution_count":3,"outputs":[{"data":{"application/vnd.jupyter.widget-view+json":{"model_id":"42de9cf45d784d94b71d04145bcaa141","version_major":2,"version_minor":0},"text/plain":"Button(description='Run the Staging Script to initialize objects', icon='gear', layout=Layout(width='50%'), stâ€¦"},"metadata":{},"output_type":"display_data"}]},{"metadata":{"trusted":false},"cell_type":"code","source":"## For viewing the various metadata files generated from the script\nout = widgets.Output(layout={'border': '1px solid black'})\nfile_open = widgets.Dropdown(\n    options=['amr_melted_analytic', 'amr_melted_raw_analytic', 'microbiome_melted_analytic', 'microbiome_melted_raw_analytic'],\n    description='file:',\n    disabled=False,\n)\n\ndef open_file(b):\n    global df\n    if file_open.value == \"amr_melted_analytic\":\n        print(\"amr_melted_analytic\")\n        df = amr_melted_analytic\n        display(df)        \n        display(out)\n    elif file_open.value == \"amr_melted_raw_analytic\":\n        print(\"amr_melted_raw_analytic\")\n        display(amr_melted_raw_analytic)\n        df = amr_melted_raw_analytic\n        display(out)\n    elif file_open.value == \"microbiome_melted_analytic\":\n        print(\"microbiome_melted_analytic\")\n        display(microbiome_melted_analytic)\n        df = microbiome_melted_analytic\n        display(out)\n    elif file_open.value == \"microbiome_melted_raw_analytic\":\n        print(\"microbiome_melted_raw_analytic\")\n        display(microbiome_melted_raw_analytic)\n        df = microbiome_melted_raw_analytic\n        display(out)\n\nopen_file_button = widgets.Button(\n    description=\"Open File\", \n    icon = 'open',\n    layout=Layout(width='70%'))\n\nopen_file_button.on_click(open_file)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Filter metadata files based on level_ID"},{"metadata":{"trusted":false},"cell_type":"code","source":"def filter_level_ID_fun(b):\n    global class_df\n    global a_select\n    global df\n    \n    if file_open.value == \"amr_melted_analytic\":\n        df = amr_melted_analytic      \n    elif file_open.value == \"amr_melted_raw_analytic\":\n        df = amr_melted_raw_analytic\n    elif file_open.value == \"microbiome_melted_analytic\":\n        df = microbiome_melted_analytic\n    elif file_open.value == \"microbiome_melted_raw_analytic\":\n        df = microbiome_melted_raw_analytic\n    \n    items = ['All']+sorted(df['Level_ID'].unique().tolist())\n\n    a_slider = widgets.IntSlider(min=0, max=len(df), step=1, value=10000)\n    a_select =  widgets.Select(options=items)\n\n    def update_select_range(*args):\n        if a_select.value=='All':\n            a_slider.max = len(df)\n        else:\n            a_slider.max = len(df[df['Level_ID']==a_select.value])\n\n    a_select.observe(update_select_range, 'value')\n\n    def view1(Level_ID='',Count=3):\n        if Level_ID=='All': return df.head(Count)\n        return df[df['Level_ID']==Level_ID].head(Count)\n\n    print(file_open.value)\n    widgets.interact(view1,Count=a_slider,Level_ID=a_select)\n\nopen_file_filter_button = widgets.Button(\n    description=\"Open File\", \n    icon = 'open',\n    layout=Layout(width='50%'))\n\nopen_file_filter_button.on_click(filter_level_ID_fun)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Print Figure"},{"metadata":{"trusted":false},"cell_type":"code","source":"def print_fig(b):\n    global class_df\n    print(\"Working...\")\n    norm = \"\"\n    if \"raw\" not in file_open.value:\n        norm = \"Normalized \"\n    \n    if a_select.value == \"All\":\n        print(\"Level_ID cannot be 'All'. Please select a different Level_ID.\")\n    else:\n        class_df = df[df['Level_ID']== str(a_select.value)]\n        import plotly.express as px\n        #df.plot(kind='bar', x='ID', y='Normalized_Count',fill='Name', mode='stack')\n        display(px.bar(x=class_df['ID'], y=class_df['Normalized_Count'] , color=class_df['Name'],\n                      labels = {\"x\" : \"Sample ID\",\n                               \"y\":str(norm + \"Read Count\"),\n                               \"color\":\"Species\"}))\n\nprint_fig_button = widgets.Button(\n    description=\"Print Plot\", \n    icon = 'open',\n    layout=Layout(width='50%'))\n\nprint_fig_button.on_click(print_fig)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Create microbiome widgets\ndirectory = widgets.Dropdown(options=os.listdir('graphs/Microbiome/'))\nimages = widgets.Dropdown(options=os.listdir('graphs/Microbiome/' + directory.value))\n\n# Updates the image options based on directory value\ndef update_images(*args):\n    images.options = os.listdir(directory.value)\n\n# Tie the image options to directory value\ndirectory.observe(update_images, 'value')\n\n# Show the images\ndef show_images(fdir, file):\n    display(Image(f'{fdir}/{file}'))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Fresh start"},{"metadata":{"trusted":false},"cell_type":"code","source":"# Creating a class to select the file for viewing\nclass figures:\n    def __init__(self):\n        self.file_view = widgets.Dropdown(\n            options=['amr_melted_analytic', 'amr_melted_raw_analytic', 'microbiome_melted_analytic', 'microbiome_melted_raw_analytic'],\n            description='file:',\n            disabled=False,\n        )\n\nrelative = figures()\ntotal = figures()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Relative Abundance"},{"metadata":{"trusted":false},"cell_type":"code","source":"# Relative Abundance Plot\n# Button function to create the file from the dropdown. Essentially save the dropdown selection and print the table\ndef print_table_rel(b): \n    global relative_df\n    global a_select\n    # exec(f\"relative_df = {relative.file_view.value}\")\n    \n    if relative.file_view.value == \"amr_melted_analytic\":\n        relative_df = amr_melted_analytic      \n    elif relative.file_view.value == \"amr_melted_raw_analytic\":\n        relative_df = amr_melted_raw_analytic\n    elif relative.file_view.value == \"microbiome_melted_analytic\":\n        relative_df = microbiome_melted_analytic\n    elif relative.file_view.value == \"microbiome_melted_raw_analytic\":\n        relative_df = microbiome_melted_raw_analytic\n    \n    \n    # items is the list of all the different Level_ID's in the given data\n    items = ['All']+sorted(relative_df['Level_ID'].unique().tolist())\n    \n    \n    # prints a select widget to select the level ID of interest\n    a_select = widgets.Select(options=items)\n    \n    # Prints the final table. the functioin allows for the table to be updated dynamically\n    def view1(Level_ID=''):\n        if Level_ID=='All': return relative_df\n        return relative_df[relative_df['Level_ID']==Level_ID]\n\n    # actually printing the table\n    widgets.interact(view1,Level_ID=a_select) \n\n\n    \n# Button info\nopen_table_rel = widgets.Button(\n    description=\"Open File for Relative Abundance Plot\", \n    icon = 'open',\n    layout=Layout(width='50%'))\n\nopen_table_rel.on_click(print_table_rel)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Create functions to graph the relative abundance plot\ndef print_rabundance_fig(b):\n    global r_class_df\n    \n    # Check to make sure that the user has selected a taxonomic level to view before running\n    try:\n        print(\"Working. . . \")\n        \n        # Updates the dataframe when new Level_ID is selected\n        r_class_df = relative_df[relative_df['Level_ID']== str(a_select.value)] \n        new_df = copy.deepcopy(r_class_df) # copies the dataframe\n\n        # These 2 lines normalize the values by ID and outputs it into a new column \"normalized_weights\"\n        group_weights = new_df.groupby('ID').aggregate(sum)\n        new_df['normalized_weights'] = new_df.apply(lambda row: row['Normalized_Count']/group_weights.loc[row['ID']][0],axis=1)\n\n        # Plots the results into relative abundance plot\n        import plotly.express as px\n        display(px.bar(x=new_df['ID'], \n                       y=new_df['normalized_weights'], \n                       color=new_df['Name'], \n                       title = str(\"Relative Abundance - \" + str(relative.file_view.value) + \": \" + str(new_df[\"Level_ID\"][0])),\n                       labels = {\"x\" : \"Sample ID\",\n                               \"y\":str(\"Normalized Read Count\"),\n                               \"color\":\"Species\"}))\n        \n    except Exception:\n        print(\"You likely have not selected a taxonomic level to view.\")\n        print(\"If you have selected a Level_ID that is not 'ALL', then another error has occured.\")\n        print(\"Contact Akhil Gupta (gupta305@umn.edu)\")\n    \n\n# Code for how the button looks\nprint_rabun_fig_button = widgets.Button(\n    description=\"Print Relative Abundance Plot\", \n    icon = 'open',\n    layout=Layout(width='70%'))\n\n\nprint_rabun_fig_button.on_click(print_rabundance_fig)\n\n\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"##### Total Abundance Plot"},{"metadata":{"trusted":false},"cell_type":"code","source":"# Total Abundance Plot\n# Button function to create the file from the dropdown. Essentially save the dropdown selection and print the table\ndef print_table_tot(b): \n    global total_df\n    global t_select\n    \n    if total.file_view.value == \"amr_melted_analytic\":\n        total_df = amr_melted_analytic      \n    elif total.file_view.value == \"amr_melted_raw_analytic\":\n        total_df = amr_melted_raw_analytic\n    elif total.file_view.value == \"microbiome_melted_analytic\":\n        total_df = microbiome_melted_analytic\n    elif total.file_view.value == \"microbiome_melted_raw_analytic\":\n        total_df = microbiome_melted_raw_analytic\n    \n    \n    # items is the list of all the different Level_ID's in the given data\n    items_tot = ['All']+sorted(total_df['Level_ID'].unique().tolist())\n    \n    \n    # prints a select widget to select the level ID of interest\n    t_select = widgets.Select(options=items_tot)\n    \n    # Prints the final table. the functioin allows for the table to be updated dynamically\n    def view2(Level_ID=''):\n        if Level_ID=='All': return total_df\n        return total_df[total_df['Level_ID']==Level_ID]\n\n    # actually printing the table\n    widgets.interact(view2,Level_ID=t_select) \n\n\n    \n# Button info\nopen_table_tot = widgets.Button(\n    description=\"Open File for Total Abundance Plot\", \n    icon = 'open',\n    layout=Layout(width='50%'))\n\nopen_table_tot.on_click(print_table_tot)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"# Create functions to graph the relative abundance plot\ndef print_tabundance_fig(b):\n    global t_class_df\n    global new_df_tot\n    \n    # Check to make sure that the user has selected a taxonomic level to view before running\n    \n    try:\n        print(\"Working. . . \")\n\n        # Updates the dataframe when new Level_ID is selected\n        t_class_df = total_df[total_df['Level_ID']== str(t_select.value)] \n        new_df_tot = copy.deepcopy(t_class_df) # copies the dataframe\n\n        # Plots the results into Total abundance plot\n        import plotly.express as px\n        display(px.bar(x=new_df_tot['ID'], \n                       y=new_df_tot['Normalized_Count'], \n                       color=new_df_tot['Name'], \n                       title = str(\"Total Abundance - \" + str(relative.file_view.value) + \": \" + str(new_df_tot[\"Level_ID\"][0])),\n                       labels = {\"x\" : \"Sample ID\",\n                               \"y\":str(\"Read Count\"),\n                               \"color\":\"Species\"}))\n    \n    except Exception:\n        print(\"You likely have not selected a taxonomic level to view.\")\n        print(\"If you have selected a Level_ID that is not 'ALL', then another error has occured.\")\n        print(\"Contact Akhil Gupta (gupta305@umn.edu)\")\n    \n\n# Code for how the button looks\nprint_tabun_fig_button = widgets.Button(\n    description=\"Print Total Abundance Plot\", \n    icon = 'open',\n    layout=Layout(width='70%'))\n\n\nprint_tabun_fig_button.on_click(print_tabundance_fig)","execution_count":null,"outputs":[]},{"metadata":{"trusted":false},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.8","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}